Documentation/update-modules-v3-file-api.md

```diff
@@ -0,0 +1,227 @@
+Update modules v3 protocol
+==========================
+
+Update modules are executables that are placed in `/usr/lib/mender/modules/v3`
+directory, where `v3` is a reference to the version of the protocol. Mender will
+look in the directory with the same version as the version of the artifact being
+processed.
+
+
+States and execution flow
+-------------------------
+
+Update modules are modelled after the same execution flow as state scripts, and
+consists of the following states:
+
+* `Download`
+* `ArtifactInstall`
+* `ArtifactReboot`
+* `ArtifactCommit`
+* `Cleanup`
+
+These all execute in the listed order, given that there are no errors. There are
+also some additional error states:
+
+* `ArtifactRollback`
+* `ArtifactRollbackReboot`
+* `ArtifactFailure`
+
+`ArtifactRollback` executes whenever:
+
+* `ArtifactInstall` has executed successfully
+* `ArtifactReboot` or `ArtifactCommit` fails
+
+`ArtifactRollbackReboot` executes whenever:
+
+* `ArtifactReboot` has executed successfully
+* `ArtifactRollback` has executed
+
+`ArtifactFailure` executes whenever:
+
+* Either of `ArtifactInstall`, `ArtifactReboot` or `ArtifactCommit` has
+  failed
+* Executes after `ArtifactRollback` and `ArtifactRollbackReboot`, if they
+  execute at all
+
+`Cleanup` executes unconditionally at the end of all the other states,
```

Comment:
Is it possible to detect which state we are cleaning up? Otherwise it would seem hard to know what to clean up?

Reply:
No, in general modules are not informed of which state they came from, which is something that's generally discouraged in state design. But the idea with `Cleanup` is that the modules would only clean up temporary files they've used (for example `rm -rf /var/tmp/modem_firmware_blobs`). "Cleanup" as in "cleaning up a failed update" is not what's intended here, that should rather happen in ArtifactRollback. Does that make sense?

I can try to make it a bit clearer that it's temporary files that should be cleaned up here, not failed updates.

Reply:
Ah, yes. Makes sense!